### PR TITLE
Move upsampling to ATen

### PR DESCRIPTION
**Pull request by colesbury**

### CODE REVIEW

File: aten/src/ATen/nn_parse.py [hunk 1]
```diff
@@ -290,6 +282,12 @@ def backward_declaration(base, thnn_functions):
 
     arguments += unique_args([output_arguments(f) for f in thnn_functions])
 
+    if 'upsample' in base['name']:
```
Comment (ezyang) [inline aten/src/ATen/nn_parse.py]: If we're adding a special case for this function, can we add a comment to the source nn.yaml file pointing out that there is a special case?

File: torch/nn/_functions/thnn/upsampling.py [hunk 2]
```diff
@@ -1,471 +0,0 @@
...
-    def backward(ctx, ggI):
-        ggO = UpsamplingLinear1d.apply(ggI, ctx.output_size, None)
-
-        return ggO, None, None
-
-
-class UpsamplingNearest2d(Function):
-
-    @staticmethod
-    def symbolic(g, input, size=None, scale_factor=None):
```
Comment (ezyang) [inline torch/nn/_functions/thnn/upsampling.py]: This needs to get ported over to torch/onnx/symbolic.py

**Review by ezyang — DISMISSED** [review]
Needs to move symbolic to `torch/onnx/symbolic.py`

File: torch/onnx/symbolic.py [hunk 3]
```diff
@@ -355,6 +355,11 @@ def replication_pad(g, input, padding):
 replication_pad3d = replication_pad
 
 
+def upsample_nearest2d(g, input, scale_factor):
+    return g.op("Upsample", input, width_scale_f=scale_factor,
+                height_scale_f=scale_factor, mode_s="nearest")
```
Comment (ezyang) [inline torch/onnx/symbolic.py]: Doesn't this also get sizes?

Comment (colesbury) [reply torch/onnx/symbolic.py]: Just `scale_factor`. If `sizes` is specified, it's converted to `scale_factor` before calling into ATen.

**Comment by vadimkantorov** [comment]
Btw `F.upsample(mode = 'bilinear')` works for downsampling as well, maybe docs or the function name should reflect this (resample?)?.

**Comment by colesbury** [comment]
@vadimkantorov, that seems reasonable, but I don't want to do it as part of this PR. I put it in #4227 if someone wants to implement it.

**Review by gchanan — APPROVED** [review]

File: torch/nn/functional.py [hunk 4]
```diff
@@ -1462,30 +1462,67 @@ def upsample(input, size=None, scale_factor=None, mode='nearest'):
...
             'nearest' | 'linear' | 'bilinear' | 'trilinear'. Default: 'nearest'
     """
+    from numbers import Integral
+    from .modules.utils import _ntuple
+
+    def _check_size_scale_factor():
+        if size is None and scale_factor is None:
+            raise ValueError('either size or scale_factor should be defined')
+        if size is not None and scale_factor is not None:
+            raise ValueError('only one of size or scale_factor should be defined')
```
[the remark this answers is not in the document]
Comment (colesbury) [reply torch/nn/functional.py]: Yes, we're a little more strict. Previously, we ignored one of the parameters if both were specified.

**Comment by gchanan** [comment]
personally I'd prefer if we had a list of all breaking changes (even if they are super minor like this).

**Pull request merged**